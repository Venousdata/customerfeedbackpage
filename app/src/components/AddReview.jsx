// this component is just a form allowing users to write and submit reviews on products

import {
  Dialog,
  Button,
  DialogTitle,
  DialogContent,
  DialogContentText,
  TextField,
  DialogActions,
} from "@material-ui/core";
import React, { useState } from "react";
import { validateProductReviewForm } from "../helpers/helpers";

const AddReview = ({ product, open, handleClose, handleSubmit }) => {
  const [name, setName] = useState("");
  const [email, setEmail] = useState("");
  const [rating, setRating] = useState(0);
  const [comment, setComment] = useState("");
  const [errors, setErrors] = useState({});

  const form = {
    name,
    email,
    rating,
    comment,
  };

  const validateAndSubmit = async (form) => {
    const errors = validateProductReviewForm(form);
    // if any errors exist --check all props of error obj
    if (errors.name || errors.email || errors.rating || errors.comment) {
      setErrors(errors);
      return;
    }

    // if validation passes, post form object to db (reviews table)
    await handleSubmit(form);
  };

  return (
    <div>
      <Dialog open={open} onClose={handleClose}>
        <DialogTitle>Add a review for {product?.name}</DialogTitle>
        <DialogContent>
          <DialogContentText>
            Fill in the fields below to submit a review!
          </DialogContentText>
          <TextField
            autoFocus
            error={errors.name ? true : false}
            margin="dense"
            label="Name"
            value={form.name}
            type="text"
            fullWidth
            onChange={(e) => {
              setName(e.target.value);
            }}
            helperText={errors.name}
          />
          <TextField
            autoFocus
            error={errors.email ? true : false}
            margin="dense"
            label="Email Address"
            type="email"
            fullWidth
            onChange={(e) => {
              setEmail(e.target.value);
            }}
            helperText={errors.email}
          />
          <TextField
            autoFocus
            error={errors.rating ? true : false}
            margin="dense"
            label="Rating"
            type="number"
            InputProps={{ inputProps: { min: "0", max: "5" } }}
            fullWidth
            defaultValue={0}
            onChange={(e) => {
              setRating(e.target.value);
            }}
            helperText={errors.rating}
          />
          <TextField
            autoFocus
            error={errors.comment ? true : false}
            margin="dense"
            label="Comment"
            multiline
            rows={5}
            fullWidth
            onChange={(e) => {
              setComment(e.target.value);
            }}
            helperText={errors.comment}
          />
        </DialogContent>
        <DialogActions>
          <Button onClick={handleClose}>Cancel</Button>
          <Button onClick={() => validateAndSubmit(form)}>Submit</Button>
        </DialogActions>
      </Dialog>
    </div>
  );
};

export default AddReview;
