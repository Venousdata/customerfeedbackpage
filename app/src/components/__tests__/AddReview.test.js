import React from "react";
import { render, screen } from "@testing-library/react";
import AddReview from "../AddReview";
import {
  mockProduct,
  mockCloseHandler,
  mockHandleSubmit,
} from "./mockData/AddReview.data";

describe("AddReview tests", () => {
  // unfinished
  it("AddReview should render and match snapshot", (done) => {
    render(
      <AddReview
        product={mockProduct}
        open={true}
        handleClose={mockCloseHandler}
        handleSubmit={mockHandleSubmit}
      />
    );
    const component = screen.getByTestId("addreview");
    expect(component.innerHTML).toMatchSnapshot();
    done();
  });
});
