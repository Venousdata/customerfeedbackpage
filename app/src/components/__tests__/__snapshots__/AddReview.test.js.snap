// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddReview tests AddReview should render and match snapshot 1`] = `"<div aria-hidden=\\"true\\"><div data-testid=\\"addreview\\"></div></div><div role=\\"presentation\\" class=\\"MuiDialog-root\\" style=\\"position: fixed; z-index: 1300; right: 0px; bottom: 0px; top: 0px; left: 0px;\\"><div class=\\"MuiBackdrop-root\\" aria-hidden=\\"true\\" style=\\"opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;\\"></div><div tabindex=\\"0\\" data-test=\\"sentinelStart\\"></div><div class=\\"MuiDialog-container MuiDialog-scrollPaper\\" style=\\"opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;\\" role=\\"none presentation\\" tabindex=\\"-1\\"><div class=\\"MuiPaper-root MuiDialog-paper MuiDialog-paperScrollPaper MuiDialog-paperWidthSm MuiPaper-elevation24 MuiPaper-rounded\\" role=\\"dialog\\"><div class=\\"MuiDialogTitle-root\\"><h2 class=\\"MuiTypography-root MuiTypography-h6\\">Add a review for fake name</h2></div><div class=\\"MuiDialogContent-root\\"><p class=\\"MuiTypography-root MuiDialogContentText-root MuiTypography-body1 MuiTypography-colorTextSecondary\\">Fill in the fields below to submit a review!</p><div class=\\"MuiFormControl-root MuiTextField-root MuiFormControl-marginDense MuiFormControl-fullWidth\\"><label class=\\"MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-marginDense\\" data-shrink=\\"false\\">Name</label><div class=\\"MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl MuiInputBase-marginDense MuiInput-marginDense\\"><input aria-invalid=\\"false\\" type=\\"text\\" class=\\"MuiInputBase-input MuiInput-input MuiInputBase-inputMarginDense MuiInput-inputMarginDense\\" value=\\"\\"></div></div><div class=\\"MuiFormControl-root MuiTextField-root MuiFormControl-marginDense MuiFormControl-fullWidth\\"><label class=\\"MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-marginDense\\" data-shrink=\\"false\\">Email Address</label><div class=\\"MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl MuiInputBase-marginDense MuiInput-marginDense\\"><input aria-invalid=\\"false\\" type=\\"email\\" class=\\"MuiInputBase-input MuiInput-input MuiInputBase-inputMarginDense MuiInput-inputMarginDense\\" value=\\"\\"></div></div><div class=\\"MuiFormControl-root MuiTextField-root MuiFormControl-marginDense MuiFormControl-fullWidth\\"><label class=\\"MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-marginDense\\" data-shrink=\\"false\\">Rating</label><div class=\\"MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl MuiInputBase-marginDense MuiInput-marginDense\\"><input aria-invalid=\\"false\\" type=\\"text\\" class=\\"MuiInputBase-input MuiInput-input MuiInputBase-inputMarginDense MuiInput-inputMarginDense\\" value=\\"\\"></div></div><div class=\\"MuiFormControl-root MuiTextField-root MuiFormControl-marginDense MuiFormControl-fullWidth\\"><label class=\\"MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-marginDense Mui-focused Mui-focused\\" data-shrink=\\"true\\">Comment</label><div class=\\"MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth Mui-focused Mui-focused MuiInputBase-formControl MuiInput-formControl MuiInputBase-multiline MuiInput-multiline MuiInputBase-marginDense MuiInput-marginDense\\"><textarea aria-invalid=\\"false\\" rows=\\"5\\" class=\\"MuiInputBase-input MuiInput-input MuiInputBase-inputMultiline MuiInput-inputMultiline MuiInputBase-inputMarginDense MuiInput-inputMarginDense\\"></textarea></div></div></div><div class=\\"MuiDialogActions-root MuiDialogActions-spacing\\"><button class=\\"MuiButtonBase-root MuiButton-root MuiButton-text\\" tabindex=\\"0\\" type=\\"button\\" data-testid=\\"cancelbutton\\"><span class=\\"MuiButton-label\\">Cancel</span><span class=\\"MuiTouchRipple-root\\"></span></button><button class=\\"MuiButtonBase-root MuiButton-root MuiButton-text\\" tabindex=\\"0\\" type=\\"button\\"><span class=\\"MuiButton-label\\">Submit</span><span class=\\"MuiTouchRipple-root\\"></span></button></div></div></div><div tabindex=\\"0\\" data-test=\\"sentinelEnd\\"></div></div>"`;
